Documentation/Evolution/UnicodeForStringProcessing.md

```diff
@@ -0,0 +1,691 @@
+
+</details>
+
+## Proposed solution
+
+TK: semantic levels, options for controlling, canonical equivalence, Unicode properties
+
+## Detailed design
+
+First, we'll discuss the options that let you control a regex's behavior, and then explore the character classes that define the your pattern.
+
+### Options
+
+Options can be enabled and disabled in two different ways: as part of [regular expression internal syntax][internals], or applied as methods when declaring a `Regex`. For example, both of these `Regex`es are declared with case insensitivity:
+
+```swift
+let regex1 = /(?i)banana/
+let regex2 = Regex {
+    "banana"
+}.ignoringCase()`
+```
+
+Note that the `ignoringCase()` is available on any type conforming to `RegexComponent`, which means that you can use the more readable option-setting interface in conjunction with regex literals:
+
+```swift
+let regex3 = /banana/.ignoringCase()
+```
+
+Calling an option-setting method like `ignoringCase(_:)` acts like wrapping the callee in an option-setting group `(?:...)`. That is, while it sets the behavior for the callee, it doesn’t override options that are applied to more specific regions. In this example, the middle `"na"` in `"banana"` matches case-sensitively, despite the outer call to `ignoringCase()`:
+
+```swift
+let regex4 = Regex {
+    "ba"
+    "na".ignoringCase(false)
+    "na"
+}
+.ignoringCase()
+	
+"banana".contains(regex4)     // true
+"BAnaNA".contains(regex4)     // true
+"BANANA".contains(regex4)     // false
+
+// Equivalent to:
+let regex5 = /(?i)ba(?-i:na)na/
+```
+
+#### Case insensitivity
+
+Regular expressions perform case sensitive comparisons by default. The `i` option or the `ignoringCase(_:)` method enables case insensitive comparison.
+
+```swift
+let str = "Café"
+	
+str.firstMatch(of: /CAFÉ/)          // nil
+str.firstMatch(of: /(?i)CAFÉ/)      // "Café"
+str.firstMatch(of: /(?i)cAfÉ/)      // "Café"
+```
+
+Case insensitive matching uses case folding to ensure that canonical equivalence continues to operate as expected.
+
+**Regular expression syntax:** `(?i)...` or `(?i:...)`
+
+**`RegexBuilder` API:**
+
+```swift
+extension RegexComponent {
+    /// Returns a regular expression that ignores casing when matching.
+    public func ignoringCase(_ ignoreCase: Bool = true) -> Regex<Output>
+}
+```
+
+#### Single line mode (`.` matches newlines)
+
+The "any" metacharacter (`.`) matches any character in a string *except* newlines by default. With the `s` option enabled, `.` matches any character including newlines.
+
+```swift
+let str = """
+    <<This string
+    uses double-angle-brackets
+    to group text.>>
+    """
+    
+str.firstMatch(of: /<<.+>>/)        // nil
+str.firstMatch(of: /(?s)<<.+>>/)    // "This string\nuses double-angle-brackets\nto group text."
+```
+
+This option also affects the behavior of `CharacterClass.any`, which is designed to match the behavior of the `.` regex literal component.
```

Comment:
I think we should conceptually think of this as being convenience API on top of more general purpose character class customizations.